### ISSUE TITLE
[CM] Add VirtualMachine support

### ISSUE THREAD

**Issue by KJlaccHoeUM9l** [Deelvin/ck#19]
At the moment, when using TVM, `GraphExecutor` is used "under hood".
However, this executor is being gradually superseded by the `VirtualMachine`. Most of the current development is done using the `VirtualMachine`, so support for `VirtualMachine` in CM needs to be added.

**Comment by KJlaccHoeUM9l** [comment]
Serialize / deserialize:
```Python
def get_tvm_executor(
    executor_type: AnyStr, mod: tvm.IRModule, params: Dict, target: AnyStr
) -> Union[
    tvm.runtime.vm.Executable,
    tvm.relay.backend.executor_factory.GraphExecutorFactoryModule,
]:
    if executor_type == "vm":
        lib = vm.compile(
            copy.deepcopy(mod),
            target,
            params=params,
        )
    elif executor_type == "graph":
        lib = relay.build(mod, target=target, params=params)
    else:
        raise ValueError(
            f"Executor type {executor_type} is unsupported. Only vm and graph types are supported"
        )
    return lib


def compile_with_statistics(
    onnx_model_path: AnyStr,
    target: AnyStr,
    tuning_logfile: AnyStr,
    executor_type: AnyStr,
    tuner_type="ansor",
) -> Union[
    tvm.runtime.vm.Executable,
    tvm.relay.backend.executor_factory.GraphExecutorFactoryModule,
]:
    model = onnx.load(onnx_model_path)
    mod, params = utils_common.get_mod_params(model)

    if tuner_type == "ansor":
        with auto_scheduler.ApplyHistoryBest(tuning_logfile):
            with tvm.transform.PassContext(
                opt_level=3,
                config={
                    "relay.backend.use_auto_scheduler": True,
                    "relay.FuseOps.max_depth": 30,
                },
            ):
                lib = get_tvm_executor(executor_type, mod, params, target)
    elif tuner_type == "meta":
        print("Use tuning files from directory:", tuning_logfile)
        database = get_json_database(tuning_logfile)

        lib = ms.relay_integration.compile_relay(
            database=database,
            mod=mod,
            target=target,
            params=params,
            backend="vm",  # "graph" by default
            opt_level=3,
        )
    else:
        raise ValueError(f"Unsupported tuner: {tuner_type}")
    return lib


def compile_virtual_machine(
    onnx_model_path: AnyStr, target: AnyStr, tuning_logfile: AnyStr
) -> tvm.runtime.vm.Executable:
    return compile_with_statistics(
        onnx_model_path,
        target,
        tuning_logfile,
        "vm",
        "meta" if "meta" in tuning_logfile else "ansor",
    )


def serialize_virtual_machine(
    vm_exec: tvm.runtime.vm.Executable, directory: AnyStr, lib_name: AnyStr
) -> None:
    path_consts = os.path.join(directory, "consts")
    vm_exec.move_late_bound_consts(path_consts, byte_limit=256)
    lib_path = os.path.join(directory, f"{lib_name}.so")
    code_path = os.path.join(directory, "vm_exec_code.ro")
    code, lib = vm_exec.save()
    lib.export_library(lib_path)
    with open(code_path, "wb") as fo:
        fo.write(code)


def deserialize_virtual_machine(
    exported_module_path: str,
    exported_consts_path: str,
    serialized_vm_exec_path: str,
) -> tvm.runtime.vm.Executable:
    mod = tvm.runtime.load_module(str(exported_module_path))

    with open(serialized_vm_exec_path, "rb") as f:
        vm_bytes = f.read()

    vm_exec = tvm.runtime.vm.Executable.load_exec(vm_bytes, mod)
    vm_exec.mod["load_late_bound_consts"](str(exported_consts_path))
    return vm_exec


def compile_graph_executor(
    onnx_model_path: AnyStr, target: AnyStr, tuning_logfile: AnyStr
) -> tvm.relay.backend.executor_factory.GraphExecutorFactoryModule:
    return compile_with_statistics(onnx_model_path, target, tuning_logfile, "graph")


def serialize_graph_executor(
    ge_exec: tvm.relay.backend.executor_factory.GraphExecutorFactoryModule,
    directory: AnyStr,
    lib_name: AnyStr,
) -> None:
    ge_exec.export_library(os.path.join(directory, f"{lib_name}.so"))

```

**Comment by KJlaccHoeUM9l** [comment]
Create VM:
```Python
from tvm.runtime import vm as runtime_vm

vm_exec = deserialize_virtual_machine(
            str(exported_module_path),
            str(exported_consts_path),
            str(serialized_vm_exec_path),
)

vm = runtime_vm.VirtualMachine(vm_exec, ctx)
```

**Comment by KJlaccHoeUM9l** [comment]
cc @Ailurus1

**Comment by Ailurus1** [comment]
Implemented in [PR#1](https://github.com/mlcommons/ck/pull/765), [PR#2](https://github.com/mlcommons/inference/pull/1395), [PR#3](https://github.com/mlcommons/ck/pull/816)
